packages/next/vitals/vitals.ts

```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from 'react'
+import { NextWebVitalsMetric } from '../pages/_app'
+
+type ReportWebVitalsCallback = (webVitals: NextWebVitalsMetric) => any
+export const webVitalsCallbacks = new Set<ReportWebVitalsCallback>()
+const metrics: NextWebVitalsMetric[] = []
+
+export function trackWebVitalMetric(metric: NextWebVitalsMetric) {
+  metrics.push(metric)
+  webVitalsCallbacks.forEach((callback) => callback(metric))
+}
+
+export function useWebVitalsReport(callback: ReportWebVitalsCallback) {
+  const callbackRef = useRef(callback)
+
+  useEffect(() => {
+    callbackRef.current = callback
+  }, [callback])
+
+  useEffect(() => {
+    if (metrics.length) {
+      metrics.forEach((metric) => callbackRef.current(metric))
+    }
+
+    webVitalsCallbacks.add(callbackRef.current)
+    return () => {
+      webVitalsCallbacks.delete(callbackRef.current)
+    }
+  }, [])
```

Comment:
Like I mentioned [here](https://github.com/vercel/next.js/pull/28769#discussion_r707125931), we have to use a closure to wrap the latest reference of the callback in a new function first, then put it inside the set. Otherwise it still calls the initial callback when new metrics are received in the future.

```suggestion
  useEffect(() => {
    const reportMetric =
      metric => callbackRef.current(metric)

    if (metrics.length) {
      metrics.forEach(reportMetric)
    }

    webVitalsCallbacks.add(reportMetric)
    return () => {
      webVitalsCallbacks.delete(reportMetric)
    }
  }, [])
```

Reply:
There's an even bigger problem too: this doesn't support [Strict Effects](https://github.com/reactwg/react-18/discussions/18) -- it'll resend every metric. I'd suggest just doing something like this:

```ts
const lastIndexRef = useRef(null)

useEffect(() => {
  if (!lastIndexRef.current) {
    lastIndexRef.current = 0
  }
  
  const reportMetric = metric => {
    callback()
    lastIndexRef.current = i
  }
  for (let i = lastIndexRef.current; i < metrics.length; i++) {
    reportMetric(metrics[i])
  }
  
  webVitalsCallbacks.add(reportMetric)
  return () => {
    webVitalsCallbacks.delete(reportMetric)
  }
}, [callback])
```

The callback itself doesn't need to be stored in a ref, since this effect is called whenever `callback` changes. Instead, we store the last metric index we sent in a ref, that way we'll only report a metric to a given component instance at most one time, regardless of the callback identity.

Reply:
You could add back the `callbackRef` optimization if you wanted to though.

Reply:
Using index sounds like a good solution, nice catch about Strict Effects, thanks!